### DISCUSSION TITLE
Enforcing nullability with single table inheritance

### DISCUSSION THREAD

**Discussion by peay** [sqlalchemy/sqlalchemy#10670] [Usage Questions]
With [single table inheritance](https://docs.sqlalchemy.org/en/20/orm/inheritance.html#single-table-inheritance), all `mapped_columns` in subclasses must be marked `nullable=True`.

```py
class Employee(Base):
    __tablename__ = "employee"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    type: Mapped[str]

    __mapper_args__ = {
        "polymorphic_on": "type",
        "polymorphic_identity": "employee",
    }


class Manager(Employee):
    manager_data: Mapped[str] = mapped_column(nullable=True)    # <------

    __mapper_args__ = {
        "polymorphic_identity": "manager",
    }
...
```

In practice, this means that although the type for `manager_data` is `str` (and not `Optional[str]`), one can still create a `Manager` instance without `manager_data`, add it to the session, and commit it to the database.

What would be the recommended way to enforce before objects get committed to the database that such columns are not `None`?

We've considered the following options:

- A `before_commit` event listener that would scan the session for such objects. Rather cumbersome to write as you need to iterate over all possible sub-classes and their columns and their type annotations. Not sure if at this point fields that are not `Optional` but can be inferred based on relationships would already have been filled in by SQLAlchemy.
- Check constraints, that would enforce this at the SQL database level. Also cumbersome to write, and can lead to huge constraints if there are dozens of sub-classes.

**Comment by zzzeek** [comment]
you could automate the production of such CHECK constraints, I'd go for that if I were ambitious, since that way you are still using server side constraints.

for Python events I would use the before_insert / before_update events here, which give you the object in question directly, rather than having to scan objects on commit.  see https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.before_insert
